tasks/cephfs/vstart_runner.py

```diff
@@ -0,0 +1,782 @@
+
+        self.exitstatus = self.returncode = self.subproc.returncode
+
+        if self.exitstatus != 0:
+            sys.stderr.write(out)
+            sys.stderr.write(err)
+
+        if self.check_status and self.exitstatus != 0:
+            raise CommandFailedError(" ".join(self.args), self.exitstatus)
+
+    @property
+    def finished(self):
+        if self.exitstatus is not None:
+            return True
+
+        if self.subproc.poll() is not None:
+            out, err = self.subproc.communicate()
+            self.stdout.write(out)
+            self.stderr.write(err)
+            self.exitstatus = self.returncode = self.subproc.returncode
+            return True
+        else:
+            return False
+
+    def kill(self):
+        log.info("kill ")
+        if self.subproc.pid and not self.finished:
+            log.info("kill: killing pid {0} ({1})".format(
+                self.subproc.pid, self.args))
+            safe_kill(self.subproc.pid)
+        else:
+            log.info("kill: already terminated ({0})".format(self.args))
+
+    @property
+    def stdin(self):
+        class FakeStdIn(object):
+            def __init__(self, mount_daemon):
+                self.mount_daemon = mount_daemon
+
+            def close(self):
+                self.mount_daemon.kill()
+
+        return FakeStdIn(self)
+
+
+class LocalRemote(object):
+    """
+    Amusingly named class to present the teuthology RemoteProcess interface when we are really
+    running things locally for vstart
+
+    Run this inside your src/ dir!
+    """
+
+    def __init__(self):
+        self.name = "local"
+        self.hostname = "localhost"
+
+    def run(self, args, check_status=True, wait=True, stdout=None, stderr=None, cwd=None, stdin=None):
+        # We don't need no stinkin' sudo
+        args = [a for a in args if a != "sudo"]
+
+        for arg in args:
+            if not isinstance(arg, basestring):
+                raise RuntimeError("Oops, can't handle arg {0} type {1}".format(
+                    arg, arg.__class__
+                ))
+
+        log.info("Running {0}".format(args))
+        subproc = subprocess.Popen(args,
+                                   stdout=subprocess.PIPE,
+                                   stderr=subprocess.PIPE,
+                                   stdin=subprocess.PIPE,
+                                   cwd=cwd)
+
+        if stdin and isinstance(stdin, basestring):
+            # Hack: writing to stdin is not deadlock-safe, but it "always" works
+            # as long as the input buffer is "small"
+            subproc.stdin.write(stdin)
+
+        proc = LocalRemoteProcess(
+            args, subproc, check_status,
+            stdout, stderr
+        )
+
+        if wait:
+            proc.wait()
+
+        return proc
+
+
+# FIXME: twiddling vstart daemons is likely to be unreliable, we should probably just let vstart
+# run RADOS and run the MDS daemons directly from the test runner
```

Comment:
Does vstart actually do anything with pid files?  stop.sh is just a killall.

The spirit of this comment is that where we want services running for the duration of the test, it would be simpler to just have them as child processes of vstart_runner, so that we can do things like sending signals and then blocking on their actual termination, rather than polling to see things die etc.